### PR TITLE
Re-implement Upload CSV to Numeric Gradeable

### LINKED ISSUE #1078: Re-implement csv upload for numeric gradeables
Functionality disabled in PR #1069 
(& clearly document required format of csv file)

### PR DESCRIPTION
**Pull request by AndrewMRPI**
It currently breaks when trying to upload really large files, such as TestExam_1_UploadBig and TestExam_1_SuperZero. I believe the ajax fail responsible is in server.js line 392. By breaks, I mean it windows alert a "[SAVE ERROR] Refresh Page" and the browser page sometimes freezes. 

Closes #1078

### CODE REVIEW

**Comment by MasterOdin** [comment]
Yeah, most likely the error is that for a page of reasonable size (let's say 100+ students, at least 3 cells), you're making 300+ ajax calls which is way too much which is then doing a ton of DOM manipulation all at once (also not good). I wonder if the better solution would be to submit an ajax call to a controller which handles uploading the file, processing it, etc. and it just returns a JSON object which is an array of all processed cells from the spreadsheet and their status ("OK", "ERROR") which the calling function would then use to color in the cells appropriately. We'd still have the problem of doing a ton of DOM manipulation at once, but that might be alright, we'd just have to test.

I'm also in favor of doing that as then we'd be able to also unit test the logic of uploading a CSV a bit easier (through PHPUnit, like I did the SubmissionController tests) which is only beneficial of not leaving all of the various edge cases to using selenium to go through the upload button.

**Comment by holzbh** [comment]
I agree with @MasterOdin, in that it's quite possible we're just starving the server out of resources by inundating it with requests through successive AJAX calls, and there's really no advantage to writing these entries one-by-one when we have the whole file. 

The comment on line 607 of server.js, `//triggers change here instad of in for loops to not start too many saves at once` hints at this problem. This is reducing the number of requests by a factor of however many cells with data there are. If there were 4 columns of data then with this fix and about 4 times as many rows of data in the .csv I imagine you would still run into this problem.

Having a single call to a controller to process the entire file at once would take care of this issue.

**Comment by holzbh** [comment]
Is there a course/gradeable these are already designed to work with, or do I need to make a new gradeable in order to test these .csv files? I didn't see one with a matching number/type of columns when I looked, but I didn't look too hard. The file names should be intuitive or there should be something that explains what the purpose/use of each file is if we're going to provide them as samples.

**Comment by AndrewMRPI** [comment]
Those were designed to work on examNumTextTest01_Title, but yeah the names might be too vague. I'll fix the errors later in the day. Thank you both for the reviews.

**Comment by AndrewMRPI** [comment]
Would first putting the data on the form with appropriate colors and then using ajax to save each cell also be acceptable?

**Comment by AndrewMRPI** [comment]
Ready for review

File: sample_files/sample_CSV/test_grades.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+assigned_dict = parse_assigned_zones()
+
+output_list = []
+file_list = ['TestExam_1']
+for file in file_list:
+    with open('%s.csv' % (file, ), 'r') as zones:
+        # Get header row contents
+        header_str = zones.readline()
+
+        line_list = zones.readlines()
```
Comment (holzbh) [inline sample_files/sample_CSV/test_grades.py]: When parsing CSV files I recommend using the built in CSV library. Right now this will have issues if any field has a comma as part of the value. (For example, comments might contain something like `"Late, had excuse"`, or zones might be something like `"HOUSTON35,10"`). I'd recommend using [Python's csv library](https://docs.python.org/3/library/csv.html) to avoid having fragile parsing. 

It'll also reduce the next few lines into one `for ____ in csvreader:` line

File: sample_files/sample_CSV/test_grades.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+                temp_array.append('%s' % (student_first_name))
+                temp_array.append('%s' % (student_last_name))
+                x = 10
+                while x < len(record):
+                    student_question_grade = float(record[x])
+                    temp_array.append('%s' % (record[x]))
+                    x+=1
+                temp_array.append('%s' % (student_grade))
+                temp_array.append('%s' % (student_assigned_zone))
+                temp_str = ','.join(temp_array)
```
Comment (holzbh) [inline sample_files/sample_CSV/test_grades.py]: Similarly, using the Python csv library means you can just ship `output_list` to `csv.writer()::writerow()` without having to worry about the separators or wrapping anything in quotation marks if needed.

File: sample_files/sample_CSV/test_grades.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+                if student_assigned_zone == 'AA':
+                    student_assigned_zone = 'UNASSIGNED'
+                if student_actual_zone == 'AA':
+                    student_actual_zone = 'UNASSIGNED'
+
+                temp_array.append('%s' % (student_rcs))
+                temp_array.append('%s' % (student_first_name))
+                temp_array.append('%s' % (student_last_name))
+                x = 10
+                while x < len(record):
```
Comment (holzbh) [inline sample_files/sample_CSV/test_grades.py]: Is `student_question_grade` used? If not, there's no reason to declare it / cast to a float.

I believe the preferred thing to do here is a list comprehension, `temp_array += [record[i] for i in range(x)]` or something to that effect.

File: sample_files/sample_CSV/test_grades.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+        # line_list = line_list[:1]
+        for index, line in enumerate(line_list):
+            line = line.strip()
+            if len(line) > 0 and 'Missing' not in line:
+                temp_array = []
+                record = line.split(',')
+                student_rcs = record[1]
+                student_name = record[0]
+                student_tmp = student_name.rsplit(' ', 1)
```
Comment (holzbh) [inline sample_files/sample_CSV/test_grades.py]: A last name can have multiple "words" in it, so don't use `rsplit(' ',1)` on real data if it needs to be accurate. There's not really a reliable way to figure out where to chop up a name if it's more than 2 parts, so it's fine for an automatic attempt to break up a name.

File: site/app/controllers/grading/SimpleGraderController.php [hunk 5]
```diff
@@ -131,4 +134,80 @@ public function save($action) {
         $this->core->getOutput()->renderJson($response);
         return $response;
     }
+
+    public function csv($action) {
```
Comment (holzbh) [inline site/app/controllers/grading/SimpleGraderController.php]: `csv` is a very generic name that doesn't adequately describe what the function does.

File: site/app/controllers/grading/SimpleGraderController.php [hunk 6]
```diff
@@ -131,4 +134,80 @@ public function save($action) {
...
+            $temp_array = explode(',', $csv_array[$i]);
+            $data_array[] = $temp_array;
+        }
+
+        foreach($users as $username) {
+            for ($j = 0; $j < $arr_length; $j++) {
+                if($username === $data_array[$j][0]) {
+                    $temp_array = array();
+                    $temp_array['username'] = $username;
+                    $num_questions = 0;
```
Comment (holzbh) [inline site/app/controllers/grading/SimpleGraderController.php]: Is `$num_questions` used anywhere else? If not, you should get rid of this line.

File: site/app/views/NavigationView.php [hunk 7]
```diff
@@ -289,6 +289,7 @@ public function showGradeables($sections_to_list) {
                  </button>
 HTML;
 						}
+
```
Comment (holzbh) [inline site/app/views/NavigationView.php]: Get rid of this newline if it's accidental. If it's not, I'd argue against having the newline, because it makes it harder to tell which if-else logic block the `else` belongs to.

File: site/public/js/server.js [hunk 8]
```diff
@@ -530,7 +529,7 @@ $(function() {
     setupNumericTextCells();
 });
 
-function setupNumericTextCells() {
+function setupNumericTextCells() {
```
Comment (holzbh) [inline site/public/js/server.js]: Get rid of the trailing space on this line

File: site/public/js/server.js [hunk 9]
```diff
@@ -584,13 +583,145 @@ function setupNumericTextCells() {
             buildUrl({'component': 'grading', 'page': 'simple', 'action': 'save_numeric'}),
             {'csrf_token': csrfToken, 'user_id': $(this).parent().parent().data("user"), 'g_id': $(this).parent().parent().data('gradeable'), 'scores': scores},
             function() {
-                $(elem).css("background-color", "#ffffff");
+                $(elem).css("background-color", "#ffffff");
```
Comment (holzbh) [inline site/public/js/server.js]: Get rid of the trailing space on this line

File: sample_files/sample_CSV/test_grades.py (hunk 4 above)
Comment (AndrewMRPI) [inline sample_files/sample_CSV/test_grades.py]: Yeah, I talked with Barb on this. We agreed that breaking up the name isn't too important since we're not using the field in the upload. So I only broke it up as the last word in a name to be the last name.

**Comment by holzbh** [comment]
It's not immediately apparent what scores I would have to have in order for sampleCSVupload_3numeric1textBad.csv to actually have bad numbers (unless I go poking around in the file). Documentation is a separate step though, so I've marked my review as approved.

What about all the csv files starting with "TestExam_1_" - are those supposed to be samples as well? If not, do they belong in the repo/in this directory?

**Comment by AndrewMRPI** [comment]
Those are just files I used for testing, I can get rid of them.

**Comment by holzbh** [comment]
Go ahead and remove them.

**Comment by bmcutler** [comment]
@holzbh do you approve?

**Comment by holzbh** [comment]
@bmcutler Yes, I approve.
